internal/service/invoice.go

```diff
@@ -265,6 +266,373 @@ func (s *invoiceService) GetInvoice(ctx context.Context, id string) (*dto.Invoic
+				proportionalCost := analyticsItem.TotalUsage.Div(totalUsageForLineItem).Mul(totalLineItemCost)
+				cost = proportionalCost.String()
+			} else {
+				cost = "0"
+			}
+
+			// Calculate percentage
+			var percentage string
+			if !totalUsageForLineItem.IsZero() {
+				pct := analyticsItem.TotalUsage.Div(totalUsageForLineItem).Mul(decimal.NewFromInt(100))
+				percentage = pct.String()
+			} else {
+				percentage = "0"
+			}
+
+			// Create usage analytics item
+			usageItem := dto.UsageAnalyticsItem{
+				Source: source,
+				Cost:   cost,
+			}
+
+			// Add optional fields
+			if !analyticsItem.TotalUsage.IsZero() {
+				usageStr := analyticsItem.TotalUsage.String()
+				usageItem.Usage = &usageStr
+			}
+
+			if percentage != "0" {
+				usageItem.Percentage = &percentage
+			}
+
+			if analyticsItem.EventCount > 0 {
+				eventCount := int(analyticsItem.EventCount)
+				usageItem.EventCount = &eventCount
+			}
+
+			lineItemUsageAnalytics = append(lineItemUsageAnalytics, usageItem)
+		}
+
+		usageAnalyticsResponse[lineItemID] = lineItemUsageAnalytics
+
+		s.Logger.Debugw("mapped usage analytics for line item",
+			"line_item_id", lineItemID,
+			"feature_id", featureID,
+			"sources_count", len(lineItemUsageAnalytics),
+			"total_usage", totalUsageForLineItem.String())
+	}
+
+	s.Logger.Infow("completed bulk analytics mapping",
+		"line_items_processed", len(usageAnalyticsResponse),
+		"total_analytics_items", s.countTotalAnalyticsItems(usageAnalyticsResponse))
+
+	return usageAnalyticsResponse, nil
+}
+
+func (s *invoiceService) CalculatePriceBreakdown(ctx context.Context, inv *dto.InvoiceResponse) (map[string][]dto.UsageAnalyticsItem, error) {
+	s.Logger.Infow("calculating price breakdown for invoice",
+		"invoice_id", inv.ID,
+		"period_start", inv.PeriodStart,
+		"period_end", inv.PeriodEnd,
+		"line_items_count", len(inv.LineItems))
+
+	// Step 1: Get the line items which are metered (usage-based)
+	usageBasedLineItems := make([]*dto.InvoiceLineItemResponse, 0)
+	for _, lineItem := range inv.LineItems {
+		if lineItem.PriceType != nil && *lineItem.PriceType == string(types.PRICE_TYPE_USAGE) {
+			usageBasedLineItems = append(usageBasedLineItems, lineItem)
+		}
+	}
+
+	s.Logger.Infow("found usage-based line items",
+		"total_line_items", len(inv.LineItems),
+		"usage_based_line_items", len(usageBasedLineItems))
+
+	if len(usageBasedLineItems) == 0 {
+		// No usage-based line items, return empty analytics
+		return make(map[string][]dto.UsageAnalyticsItem), nil
+	}
+
+	// OPTIMIZED: Use single ClickHouse call to get all analytics data grouped by source and feature_id
+	return s.getBulkUsageAnalyticsForInvoice(ctx, usageBasedLineItems, inv)
+}
+
+// Helper function to count total analytics items
+func (s *invoiceService) countTotalAnalyticsItems(analytics map[string][]dto.UsageAnalyticsItem) int {
+	total := 0
+	for _, items := range analytics {
+		total += len(items)
+	}
+	return total
+}
+
```

Comment:
For future - we use lodash to avoid writing these common wrappers
can check if there is a matching function here https://github.com/samber/lo?tab=readme-ov-file#count